dnf/module/repo_module_dict.py

```diff
@@ -0,0 +1,462 @@
+            module_version = self.find_module_version(module_version.name, module_version.stream,
+                                                      preferred_version)
+        return module_version
+
+    def get_preferred_versions(self, pkg_specs):
+        preferred_versions = {}
+        for pkg_spec in pkg_specs:
+            subj = ModuleSubject(pkg_spec)
+            module_version, nsvcap = subj.find_module_version(self)
+
+            if not module_version:
+                continue
+
+            key = "{}:{}".format(module_version.name, module_version.stream)
+            versions = preferred_versions.setdefault(key, PreferredModuleVersion())
+            versions.moduleversion_nsvcap_pkgspec.append([module_version, nsvcap, pkg_spec])
+            if versions.preferred_version < module_version.version:
+                versions.preferred_version = module_version.version
+                versions.reason = pkg_spec
+
+        return preferred_versions
+
+    def upgrade(self, pkg_specs):
+        for pkg_spec in pkg_specs:
+            subj = ModuleSubject(pkg_spec)
+            module_version, nsvcap = subj.find_module_version(self)
+
+            if module_version.repo_module.conf.locked:
+                continue
+
+            conf = self[nsvcap.name].conf
+            if conf:
+                installed_profiles = conf.profiles
+            else:
+                installed_profiles = []
+            if nsvcap.profile:
+                if nsvcap.profile not in installed_profiles:
+                    raise Error(module_errors[PROFILE_NOT_INSTALLED].format(pkg_spec))
+                profiles = [nsvcap.profile]
+            else:
+                profiles = installed_profiles
+
+            pkg_specs.remove(pkg_spec)
+            module_version.upgrade(profiles)
+
+        return pkg_specs
+
+    def upgrade_all(self):
+        modules = []
+        for module_name, repo_module in self.items():
+            if not repo_module.conf:
+                continue
+            if not repo_module.conf.enabled:
+                continue
+            modules.append(module_name)
+        modules.sort()
+        self.upgrade(modules)
+
+    def remove(self, pkg_specs):
+        for pkg_spec in pkg_specs:
+            subj = ModuleSubject(pkg_spec)
+            module_version, nsvcap = subj.find_module_version(self)
+
+            conf = self[nsvcap.name].conf
+            if conf:
+                installed_profiles = conf.profiles
+            else:
+                installed_profiles = []
+            if nsvcap.profile:
+                if nsvcap.profile not in installed_profiles:
+                    raise Error(module_errors[PROFILE_NOT_INSTALLED].format(pkg_spec))
+                profiles = [nsvcap.profile]
+            else:
+                profiles = installed_profiles
+
+            module_version.remove(profiles)
+
+    def load_modules(self, repo):
```

Comment:
whats a difference between load_modules and read_all_modules and read_all_module_defaults